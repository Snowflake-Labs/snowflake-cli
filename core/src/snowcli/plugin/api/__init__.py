from dataclasses import dataclass
from typing import List

import pluggy
from typer import Typer

plugin_hook_spec = pluggy.HookspecMarker("snowcli.plugin")
plugin_hook_impl = pluggy.HookimplMarker("snowcli.plugin")


@dataclass
class PluginCommandGroupPath:
    path_segments: List[str]


@dataclass
class PluginCommandGroupSpec:
    path: PluginCommandGroupPath
    command_group: Typer
