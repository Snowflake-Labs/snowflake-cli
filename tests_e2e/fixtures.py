import shutil
from contextlib import contextmanager
from pathlib import Path

import pytest


@pytest.fixture
def project_directory(temp_dir, test_root_path):
    @contextmanager
    def _temporary_project_directory(project_name):
        test_data_file = test_root_path / "test_data" / project_name
        shutil.copytree(test_data_file, temp_dir, dirs_exist_ok=True)
        yield Path(temp_dir)

    return _temporary_project_directory
