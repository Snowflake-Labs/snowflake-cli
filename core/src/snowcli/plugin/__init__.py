from dataclasses import dataclass
from enum import Enum
from typing import List, Optional

from snowcli.plugin.api import PluginCommandGroupSpec


class PluginLoadingMode(Enum):
    ALL_INSTALLED_PLUGINS = 1
    ONLY_ENABLED_PLUGINS = 2


@dataclass
class PluginPackageInfo:
    package_name: str
    package_version: str
    dependencies: List[str]


@dataclass
class LoadedPlugin:
    plugin_name: str
    is_enabled: bool
    plugin_package_info: PluginPackageInfo
    command_group_spec: PluginCommandGroupSpec


@dataclass
class PluginAfterVerification:
    plugin: LoadedPlugin
    verification_error: Optional[Exception]

    def is_ok(self) -> bool:
        return self.verification_error is None

    def error_str(self) -> Optional[str]:
        error = self.verification_error
        if error:
            return f"{error.__class__.__name__} - {error.__str__()}"
        else:
            return None
