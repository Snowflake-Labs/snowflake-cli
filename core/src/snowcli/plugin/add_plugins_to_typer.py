import logging
from typing import List

import click
import typer
from typer.core import TyperGroup

from snowcli.cli.common.snow_cli_global_context import global_context_copy
from snowcli.plugin import LoadedPlugin
from snowcli.plugin.api import PluginCommandGroupSpec, PluginCommandGroupPath

log = logging.getLogger(__name__)


class AddPluginsToTyper:
    def __init__(self, plugins: List[LoadedPlugin]):
        self._plugins = plugins
        self._main_typer_command_group = (
            self._get_main_typer_command_group_from_click_context()
        )

    def __call__(self, *args, **kwargs):
        for plugin in self._plugins:
            try:
                self._add_plugin_command_group(plugin.command_group_spec)
            except Exception as ex:
                log.error(
                    f"Cannot register command group from plugin [{plugin.plugin_name}]: {ex.__str__()}",
                    exc_info=global_context_copy().debug,
                )

    @staticmethod
    def _get_main_typer_command_group_from_click_context() -> TyperGroup:
        main_typer_command_group = click.get_current_context().command
        if isinstance(main_typer_command_group, TyperGroup):
            return main_typer_command_group
        else:
            raise RuntimeError(
                "Invalid main top-level command type. It should be a TyperGroup but it is not."
            )

    def _add_plugin_command_group(
        self,
        command_group_spec: PluginCommandGroupSpec,
    ) -> None:
        parent_group = self._find_typer_group_at_path(
            current_level_group=self._main_typer_command_group,
            remaining_path_segments=command_group_spec.path.path_segments,
            full_plugin_path=command_group_spec.path,
        )
        plugin_group = typer.main.get_group(command_group_spec.command_group)
        parent_group.add_command(plugin_group)

    def _find_typer_group_at_path(
        self,
        current_level_group: TyperGroup,
        remaining_path_segments: List[str],
        full_plugin_path: PluginCommandGroupPath,
    ) -> TyperGroup:
        if remaining_path_segments:
            expected_name = remaining_path_segments[0]
            matching_subgroups = [
                subgroup
                for subgroup in current_level_group.commands.values()
                if isinstance(subgroup, TyperGroup) and subgroup.name == expected_name
            ]
            if matching_subgroups:
                return self._find_typer_group_at_path(
                    current_level_group=matching_subgroups[0],
                    remaining_path_segments=remaining_path_segments[1:],
                    full_plugin_path=full_plugin_path,
                )
            else:
                raise RuntimeError(
                    f"Invalid plugin path [{'/'.join(full_plugin_path.path_segments)}]. "
                    f"Command group [{expected_name}] does not exist."
                )
        else:
            return current_level_group
