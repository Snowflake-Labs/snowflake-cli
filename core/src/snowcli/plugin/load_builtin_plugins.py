import logging
from typing import List

import pluggy

import snowcli.cli.stage.manager
from snowcli.cli.common.snow_cli_global_context import global_context_copy
from snowcli.cli.stage.stage_builtin_plugin_hook_impl import plugin_spec
from snowcli.plugin import BuiltInLoadedPlugin
from snowcli.plugin.api import plugin_hook_specs

log = logging.getLogger(__name__)


class LoadBuiltInPlugins:

    _builtin_plugins = {"stage": snowcli.cli.stage.stage_builtin_plugin_hook_impl}

    def __init__(self):
        plugin_manager = pluggy.PluginManager("snowcli.plugin")
        plugin_manager.add_hookspecs(plugin_hook_specs)
        self._plugin_manager = plugin_manager

    def __call__(self, *args, **kwargs) -> List[BuiltInLoadedPlugin]:
        self._load_plugins_code_to_manager()
        return self._extract_info_and_specifications_of_loaded_plugins()

    def _load_plugins_code_to_manager(self) -> None:
        for (plugin_name, plugin) in self._builtin_plugins.items():
            self._plugin_manager.register(plugin=plugin, name=plugin_name)

    def _extract_info_and_specifications_of_loaded_plugins(
        self,
    ) -> List[BuiltInLoadedPlugin]:
        plugins = []
        for (plugin_name, plugin) in self._plugin_manager.list_name_plugin():
            try:
                plugins.append(
                    BuiltInLoadedPlugin(
                        plugin_name=plugin_name,
                        plugin_spec=plugin.plugin_spec(),
                    )
                )
            except Exception as ex:
                log.error(
                    msg=f"Cannot load info about built-in plugin [{plugin_name}]: {ex.__str__()}",
                    exc_info=global_context_copy().debug,
                )
        return plugins
