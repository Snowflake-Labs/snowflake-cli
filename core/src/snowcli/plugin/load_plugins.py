import logging
from typing import List

import pluggy

from snowcli.cli.common.snow_cli_global_context import global_context_copy
from snowcli.config import cli_config
from snowcli.plugin import PluginLoadingMode, LoadedPlugin, PluginPackageInfo
from snowcli.plugin.api import plugin_hook_specs

log = logging.getLogger(__name__)


class LoadPlugins:
    def __init__(self, plugin_loading_mode: PluginLoadingMode):
        plugin_manager = pluggy.PluginManager("snowcli.plugin")
        plugin_manager.add_hookspecs(plugin_hook_specs)
        self._plugin_manager = plugin_manager
        self._plugin_loading_mode = plugin_loading_mode
        self._enabled_plugins = cli_config.get_enabled_plugins()

    def __call__(self, *args, **kwargs) -> List[LoadedPlugin]:
        self._load_plugins_code_to_manager()
        return self._extract_info_and_specifications_of_loaded_plugins()

    def _load_plugins_code_to_manager(self) -> None:
        {
            PluginLoadingMode.ALL_INSTALLED_PLUGINS: lambda: self._load_code_of_all_installed_plugins(),
            PluginLoadingMode.ONLY_ENABLED_PLUGINS: lambda: self._load_code_of_enabled_plugins(),
        }[self._plugin_loading_mode]()

    def _load_code_of_all_installed_plugins(self) -> None:
        self._plugin_manager.load_setuptools_entrypoints(group="snowcli.plugin")

    def _load_code_of_enabled_plugins(self) -> None:
        for plugin_name in self._enabled_plugins:
            try:
                self._plugin_manager.load_setuptools_entrypoints(
                    group="snowcli.plugin", name=plugin_name
                )
            except Exception as ex:
                log.error(
                    msg=f"Cannot load code of enabled plugin [{plugin_name}]: {ex.__str__()}",
                    exc_info=global_context_copy().debug,
                )

    def _extract_info_and_specifications_of_loaded_plugins(self) -> List[LoadedPlugin]:
        plugins_to_package_info = {
            plugin: self._extract_package_info(distinfo)
            for (plugin, distinfo) in self._plugin_manager.list_plugin_distinfo()
        }
        plugins = []
        for (plugin_name, plugin) in self._plugin_manager.list_name_plugin():
            try:
                plugins.append(
                    LoadedPlugin(
                        plugin_name=plugin_name,
                        is_enabled=plugin_name in self._enabled_plugins,
                        plugin_package_info=plugins_to_package_info[plugin],
                        command_group_spec=plugin.plugin_command_group_spec(),
                    )
                )
            except Exception as ex:
                log.error(
                    msg=f"Cannot load info about plugin [{plugin_name}]: {ex.__str__()}",
                    exc_info=global_context_copy().debug,
                )
        return plugins

    @staticmethod
    def _extract_package_info(distinfo) -> PluginPackageInfo:
        dependencies = distinfo.__getattr__("requires")
        return PluginPackageInfo(
            package_name=distinfo.__getattr__("name"),
            package_version=distinfo.__getattr__("version"),
            dependencies=dependencies,
        )
