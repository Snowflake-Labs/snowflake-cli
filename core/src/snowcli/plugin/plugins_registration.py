import logging
from typing import List

from snowcli.plugin import (
    ExternalPluginsLoadingMode,
    ExternalLoadedPlugin,
    LoadedPlugin,
)
from snowcli.plugin.add_plugins_to_typer import AddPluginsToTyper
from snowcli.plugin.load_builtin_plugins import LoadBuiltInPlugins
from snowcli.plugin.load_external_plugins import LoadExternalPlugins
from snowcli.plugin.verify_external_plugins import VerifyExternalPlugins

log = logging.getLogger(__name__)


def _cast_to_loaded_plugin(plugin: LoadedPlugin) -> LoadedPlugin:
    return plugin


def load_and_register_plugins_in_typer(
    external_plugins_loading_mode: ExternalPluginsLoadingMode,
):
    loaded_builtin_plugins = LoadBuiltInPlugins()()
    loaded_external_plugins = LoadExternalPlugins(external_plugins_loading_mode)()
    verified_external_plugins: List[ExternalLoadedPlugin] = []
    for plugin_verification in VerifyExternalPlugins(loaded_external_plugins)():
        if plugin_verification.is_ok():
            verified_external_plugins.append(plugin_verification.plugin)
        else:
            log.error(
                f"Cannot register external plugin [{plugin_verification.plugin.plugin_name}] "
                f"because of verification error: {plugin_verification.error_str()}"
            )
    all_plugins_to_register = [
        _cast_to_loaded_plugin(p) for p in loaded_builtin_plugins
    ] + [_cast_to_loaded_plugin(p) for p in verified_external_plugins]
    AddPluginsToTyper(all_plugins_to_register)()
