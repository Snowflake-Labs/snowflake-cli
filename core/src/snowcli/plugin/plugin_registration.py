import logging
from typing import List

from snowcli.plugin import PluginLoadingMode, LoadedPlugin
from snowcli.plugin.add_plugins_to_typer import AddPluginsToTyper
from snowcli.plugin.load_plugins import LoadPlugins
from snowcli.plugin.verify_plugins import VerifyPlugins

log = logging.getLogger(__name__)


def load_and_register_plugins_in_typer(plugin_loading_mode: PluginLoadingMode):
    loaded_plugins = LoadPlugins(plugin_loading_mode)()
    verified_plugins: List[LoadedPlugin] = []
    for plugin_verification in VerifyPlugins(loaded_plugins)():
        if plugin_verification.is_ok():
            verified_plugins.append(plugin_verification.plugin)
        else:
            log.error(
                f"Cannot register plugin [{plugin_verification.plugin.plugin_name}] "
                f"because of verification error: {plugin_verification.error_str()}"
            )
    AddPluginsToTyper(verified_plugins)()
