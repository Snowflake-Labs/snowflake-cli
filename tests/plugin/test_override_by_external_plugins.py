from textwrap import dedent

from tests.plugin.fixtures import install_plugin  # noqa


def test_override_build_in_commands(runner, test_root_path, install_plugin, caplog):
    install_plugin("override_build_in_command")

    config_path = (
        test_root_path / "test_data" / "configs" / "override_plugin_config.toml"
    )

    result = runner.invoke(["--config-file", config_path, "connection", "list"])

    assert (
        caplog.messages[0]
        == "Cannot register plugin [override]: Cannot add command [snow connection list] because it already exists."
    )
    assert result.output == dedent(
        """Outside command code
+---------------------------------------+
| connection_name | parameters          |
|-----------------+---------------------|
| test            | {'account': 'test'} |
+---------------------------------------+
    """
    )


def test_disabled_plugin_is_not_executed(
    runner, test_root_path, install_plugin, caplog
):
    install_plugin("override_build_in_command")
    config_path = (
        test_root_path
        / "test_data"
        / "configs"
        / "disabled_override_plugin_config.toml"
    )

    result = runner.invoke(["--config-file", config_path, "connection", "list"])

    assert len(caplog.messages) == 0
    assert result.output == dedent(
        """+---------------------------------------+
| connection_name | parameters          |
|-----------------+---------------------|
| test            | {'account': 'test'} |
+---------------------------------------+
    """
    )
