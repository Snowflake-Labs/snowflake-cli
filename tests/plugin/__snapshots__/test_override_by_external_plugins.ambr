# serializer version: 1
# name: test_corrupted_config_raises_human_friendly_error[[corrupted0]
  '''
  ╭─ Error ──────────────────────────────────────────────────────────────────────╮
  │ Configuration file seems to be corrupted. Unexpected end of file at line 1   │
  │ col 10                                                                       │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  '''
# ---
# name: test_corrupted_config_raises_human_friendly_error[[corrupted1]
  '''
  ╭─ Error ──────────────────────────────────────────────────────────────────────╮
  │ Configuration file seems to be corrupted. Unexpected end of file at line 1   │
  │ col 10                                                                       │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  '''
# ---
